### PR TITLE
Electronic Constraints

### PR DESCRIPTION
**Pull request by vanderhe**
Enables to impose constraints on the electronic ground state.

- [x] general CDFTB infrastructure
- [x] available constraints: atomic Mulliken populations
- [x] atomic gradients/stress tensor
- [x] re-check analytic against numerical gradients of artificial potential
- [x] documentation
- [x] possible barriers for unsupported extensions
- [x] wait for #1226
- [x] wait for #1222
- [x] regression tests
- [x] re-check atomic gradients/stress tensor

Caveats of current implementation:
- SCC-DFTB Hamiltonian re-build in every micro-iteration, although not necessary
- LC-DFTB Hamiltonian re-build in every micro-iteration, although not necessary
- allows for impossible targets, e.g. those that would violate the total number of electrons in the system

Adding constraints, e.g. non-colinear atomic shell spin, atomic shell orbital angular momentum, atomic shell projection of orbital angular momentum, atomic shell total angular momentum, atomic shell projection of angular momentum, ..., is straightforward.

### CODE REVIEW

**Comment by bhourahine** [comment]
@vanderhe you might want to think about the Co-authored-by: tag for commits due to the history of some of the code.

**Comment by vanderhe** [comment]
> @vanderhe you might want to think about the Co-authored-by: tag for commits due to the history of some of the code.

@bhourahine Sure thing, is there anyone else we need to consider?

**Comment by vanderhe** [comment]
> Just me. Ps, the forces are just Helmann-Feynman terms from the free energy expression, so just add V as an external potential. Think I gave the expressions in on or other of my constraints papers.

Yes, the forces are working already but I experienced problems with the stress tensor (which seems to be a trivial mistake on my side because not even the one of conventional, unconstrained DFTB matches...). I will further investigate tomorrow.

**Comment by bhourahine** [comment]
The fire optimiser fix should probably be it's own separate PR.

**Comment by vanderhe** [comment]
> The fire optimiser fix should probably be it's own separate PR.

Yes, I will separate out all the optimizer related things, like introducing reset() functionality for the new ones. This is just for testing purposes right now.

**Comment by bhourahine** [comment]
You probably should print out the constraint functional energy separately from the other internal energy quantities.
I'm not seeing why several of the constraint + method combinations are blocked (atom/region charges should be OK with most things: MPI, +U, nSpin>2, L.S, shell resolved, and even xTB). Enabling nSpin>2 with debug flags up the NEf evaluation (line 1343) from commit 6ff7c97ef. This line is active(?), and broken for nSpin>2, for systems with fixed total numbers of electrons.

**Comment by vanderhe** [comment]
> You probably should print out the constraint functional energy separately from the other internal energy quantities.

The energy contribution due to the constraints is now printed out during the micro-iterations.

> I'm not seeing why several of the constraint + method combinations are blocked (atom/region charges should be OK with most things: MPI, +U, nSpin>2, L.S, shell resolved, and even xTB).

I disabled every extension where I was not absolutely sure that it works out of the box. Aforementioned ones are re-enabled by now.

> Enabling nSpin>2 with debug flags up the NEf evaluation (line 1343) from commit [6ff7c97](https://github.com/dftbplus/dftbplus/commit/6ff7c97efdd7f6d05bcf62597b6a099958a119fc). This line is active(?), and broken for nSpin>2, for systems with fixed total numbers of electrons.

I have no experience with nSpin>2 calculations. Do you already have a fix at hand that you/I can apply?

**Comment by vanderhe** [comment]
@aradi Default optimizer switched to FIRE, brief explanation added to manual and behavior of `ConvergentConstrOnly` keyword adapted as discussed. Feel free to have a look at it again.

**Comment by bhourahine** [comment]
@vanderhe can you rebase to fix the conflict and add something to the CHANGELOG.rst? I'll try and have a look at the UX in the next day or so.

**Comment by vanderhe** [comment]
> @vanderhe can you rebase to fix the conflict and add something to the CHANGELOG.rst? I'll try and have a look at the UX in the next day or so.

@bhourahine Sure, will do in a sec.

**Comment by bhourahine** [comment]
@vanderhe it would be useful to print out the free energy functional somewhere, W = E + Vc (q - Nc). But also the derivative of W with respect to the constraint value Nc, i.e. -Vc. Possibly either in detailed.out or in it's own file.

**Comment by aradi** [comment]
OK, I think, we should make the input (and the offered features) more flexible. In order not to mess up this PR, I've created #1234 where we discuss this. I'll review the rest nevertheless.

File: src/dftbp/dftb/elecconstraints.F90 [hunk 1]
```diff
@@ -7,428 +7,396 @@
...
+      call getConstrainEnergyAndPotQ(deltaW, dWdVc(iConstr), this%Vc(iConstr), this%Nc(iConstr),&
+          & this%wAt(iConstr)%data, this%wAtOrb(iConstr)%data, this%wAtSpin(iConstr)%data,&
+          & qq)
     end do
 
-    Sshell = sum(qIn(orb%posShell(conSh,species(conAt)): &
-        & orb%posShell(conSh+1,species(conAt))-1,conAt,2:4),dim=1)
+    ! Optimizers set up to minimize, therefore sign change in total energy and gradients
+    ! call next(this%potOpt, -(energy + deltaW), -dWdVc, this%Vc, tConverged)
+    call this%potOpt%step(energy + deltaW, -dWdVc, potDisplace)
```
Comment (aradi) [inline src/dftbp/dftb/elecconstraints.F90:273]: I find the comment (and the commented out code) confusing, as it is the opposite (for the energy) as actually done in the line below.

File: src/dftbp/dftb/elecconstraints.F90 [hunk 2]
```diff
@@ -7,428 +7,396 @@
...
 
-    do iSpin = 2, nSpin
-      do iOrb = orb%posShell(conSh,species(conAt)), &
-          & orb%posShell(conSh+1,species(conAt))-1
-        shift(iOrb,iOrb,conAt,iSpin) = shift(iOrb,iOrb,conAt,iSpin) &
-            & + w * vecNorm(iSpin-1)
+    do iSpin = 1, nSpin
+      do iW = 1, size(wAt)
+        shift(wOrb(iW), wOrb(iW), wAt(iW), iSpin) = shift(wOrb(iW), wOrb(iW), wAt(iW), iSpin)&
+            & + Vc * wSp(iW, iSpin)
```
Comment (aradi) [inline src/dftbp/dftb/elecconstraints.F90:396]: Don't we miss the overlap here?

File: src/dftbp/dftb/elecconstraints.F90 [hunk 3]
```diff
@@ -7,428 +7,396 @@
...
-    !> direction of spin
-    real(dp), intent(in) :: vec(3)
+contains
 
-    integer :: iOrb, nSpin, iSpin
-    real(dp) :: Sshell(3), W, vecNorm(3)
 
-    nSpin = size(shift,dim=4)
+  !> General entry point to read constraint on the electronic ground state.
+  subroutine readElecConstraintInput(node, geo, input, tSpinPol)
```
[the remark this answers is not in the document]
Comment (vanderhe) [reply src/dftbp/dftb/elecconstraints.F90:97]: I went with `isSpinPol` to be more consistent with other parts of the codebase.

File: src/dftbp/dftb/elecconstraints.F90 (hunk 1 above)
Comment (vanderhe) [reply src/dftbp/dftb/elecconstraints.F90:273]: Artifact of old version, removed in latest iteration.

File: src/dftbp/dftb/elecconstraints.F90 (hunk 2 above)
Comment (vanderhe) [reply src/dftbp/dftb/elecconstraints.F90:424]: I will mark this as resolved, since the error seems to be located in the paper, not the implementation.

**Comment by bhourahine** [comment]
@vanderhe you seem to have some suggestions in a comment from May 18th from @aradi . All look simple, so any reason not to incorporate?

**Review by bhourahine — COMMENTED** [review]
The utils/numcheck/calcderivs script gives a substantial mismatch for the forces for the non-scc H2O example. My guess is the wrong energy is being differentiated (unless there is a bug in the forces somewhere).

File: src/dftbp/dftbplus/main.F90 [hunk 4]
```diff
@@ -746,6 +749,12 @@ subroutine processPotentials(env, this, iSccIter, updateScc, q, qBlock, qiBlock)
 
     end if
 
+    if (allocated(this%elecConstrain)) then
+      constrShift(:,:,:,:) = 0.0_dp
+      call this%elecConstrain%getConstrainShift(constrShift)
+      this%potential%intBlock = this%potential%intBlock + constrShift
```
Comment (bhourahine) [inline src/dftbp/dftbplus/main.F90:755]:
```suggestion
      this%potential%intBlock(:,:,:,:) = this%potential%intBlock + constrShift
```

File: src/dftbp/dftbplus/main.F90 [hunk 5]
```diff
@@ -746,6 +749,12 @@ subroutine processPotentials(env, this, iSccIter, updateScc, q, qBlock, qiBlock)
...
+    end if
+
     ! All potentials are added up into intBlock
     this%potential%intBlock = this%potential%intBlock + this%potential%extBlock
```
Comment (bhourahine) [inline src/dftbp/dftbplus/main.F90:759]:
```suggestion
    this%potential%intBlock(:,:,:,:) = this%potential%intBlock + this%potential%extBlock
```

**Review by bhourahine — COMMENTED** [review]
Couple of left assignment dimensionings missing

**Comment by vanderhe** [comment]
> @vanderhe you seem to have some suggestions in a comment from May 18th from @aradi . All look simple, so any reason not to incorporate?

Just a mixture of having overlooked hidden conversations and forgotten to resolve applied ones; corrected in next version to be uploaded.

**Review by bhourahine — APPROVED** [review]
Ready to merge, but will need follow up PRs for #1234 and various other extensions.

**Pull request merged**